lib/rules/no-useless-assertions.ts

```diff
@@ -0,0 +1,319 @@
+        "The negative {{kind}} '{{assertion}}' will always {{acceptOrReject}}.",
+}
+
+export default createRule("no-useless-assertions", {
+    meta: {
+        docs: {
+            description:
+                "disallow assertions that are known to always accept (or reject)",
+            // TODO Switch to recommended in the major version.
+            // recommended: true,
+            recommended: false,
+        },
+        schema: [],
+        messages,
+        type: "problem",
+    },
+    create(context) {
+        const sourceCode = context.getSourceCode()
+
+        /**
+         * Create visitor
+         * @param node
+         */
+        function createVisitor(
+            node: Expression,
+            _pattern: string,
+            flagsStr: string,
+        ): RegExpVisitor.Handlers {
+            const flags = parseFlags(flagsStr)
+            const flagsWithoutDotAll = parseFlags(flagsStr.replace(/s/g, ""))
+
+            /** Report */
+            function report(
+                assertion: Assertion,
+                messageId: keyof typeof messages,
+                data: Record<string, string>,
+            ) {
+                context.report({
+                    node,
+                    loc: getRegexpLocation(sourceCode, node, assertion),
+                    messageId,
+                    data: {
+                        assertion: assertion.raw,
+                        ...data,
+                    },
+                })
+            }
+
+            /**
+             * Verify for `^` or `$`
+             */
+            function verifyStartOrEnd(assertion: EdgeAssertion): void {
+                // Note: /^/ is the same as /(?<!.)/s and /^/m is the same as /(?<!.)/
+                // Note: /$/ is the same as /(?!.)/s and /$/m is the same as /(?!.)/
+
+                // get the "next" character
+                const direction = getMatchingDirectionFromAssertionKind(
+                    assertion.kind,
+                )
+                const next = getFirstCharAfter(assertion, direction, flags)
+
+                const followedOrPreceded =
+                    assertion.kind === "end" ? "followed" : "preceded"
+
+                if (!next.edge) {
+                    // there is always some character of `node`
+
+                    if (!flags.multiline) {
+                        // since the m flag isn't present any character will result in trivial rejection
+                        report(assertion, "alwaysRejectByChar", {
+                            followedOrPreceded,
+                        })
+                    } else {
+                        // only if the character is a sub set of /./, will the assertion trivially reject
+
+                        // with this little flag hack, we can easily create the dot set.
+
+                        const dot = JS.createCharSet(
+                            [{ kind: "any" }],
+                            flagsWithoutDotAll,
+                        )
```

Comment:
You can use the `Chars` namespace from `regexp-ast-analysis` here.

```suggestion
                        const dot = Chars.lineTerminator(flags).negate()
```